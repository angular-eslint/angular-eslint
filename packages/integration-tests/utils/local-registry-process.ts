/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { ChildProcess } from 'child_process';
import { spawn } from 'child_process';
import execa from 'execa';
import { join } from 'path';
import kill from 'tree-kill';

export const FIXTURES_DIR = join(__dirname, '../fixtures/');

// Used to ensure all the time-consuming setup steps for fixtures do not cause jest to time out
export const LONG_TIMEOUT_MS = 600000; // 10 mins

let localRegistryProcess: ChildProcess;

const VERSION = `9999.0.1-local-integration-tests`;

const cwd = process.cwd();

process.env.npm_config_registry = `http://localhost:4872/`;
process.env.YARN_REGISTRY = process.env.npm_config_registry;

export async function spawnLocalRegistry(): Promise<void> {
  localRegistryProcess = spawn('npx', [
    'verdaccio',
    '--config',
    './local-registry/config.yml',
    '--listen',
    '4872',
  ]);

  const collectedOutput: string[] = [];
  let resolvedOrRejected = false;

  setTimeout(() => {
    if (!resolvedOrRejected) {
      console.error(`Failed to start the npm registry`);
      console.error(collectedOutput.join(''));
      cleanUp(1);
    }
  }, 10000);

  await new Promise((res, rej) => {
    localRegistryProcess.stdout!.pipe(process.stdout);
    localRegistryProcess.stdout!.on('data', (data: Buffer) => {
      collectedOutput.push(data.toString());
      // wait for local-registry to come online
      if (data.includes('http address')) {
        resolvedOrRejected = true;
        res(undefined);
      }
    });
    localRegistryProcess.stderr!.pipe(process.stderr);
    localRegistryProcess.on('error', (err: Error) => {
      console.error(collectedOutput.join(''));
      resolvedOrRejected = true;
      rej(err);
    });
  });
}

export function cleanUp(code: number): void {
  process.chdir(cwd);

  // try and terminate everything first
  try {
    if (!process.env.CI) {
      kill(0);
    }
    // eslint-disable-next-line no-empty
  } catch {}
  try {
    if (localRegistryProcess) localRegistryProcess.kill();
    // eslint-disable-next-line no-empty
  } catch {}
  // try killing everything after in case something hasn't terminated
  try {
    if (!process.env.CI) {
      kill(0, 'SIGKILL');
    }
    // eslint-disable-next-line no-empty
  } catch {}
  try {
    if (localRegistryProcess) localRegistryProcess.kill('SIGKILL');
    // eslint-disable-next-line no-empty
  } catch {}

  process.exit(code);
}

export async function publishPackagesToVerdaccio(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('./publish-to-verdaccio.sh', [VERSION]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNpmInstall(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npm', ['install']);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runYarnInstall(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('yarn', ['install']);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgAdd(): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npx', [
    'ng',
    'add',
    `@angular-eslint/schematics@${VERSION}`,
  ]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgNew(
  workspaceName: string,
  createApplication = true,
): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const ngNewArgs = [
    `--strict=true`,
    `--package-manager=npm`,
    `--interactive=false`,
  ];

  if (!createApplication) {
    ngNewArgs.push(`--create-application=false`);
  }

  const subprocess = execa('../../../node_modules/.bin/ng', [
    'new',
    ...ngNewArgs,
    workspaceName,
  ]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgGenerate(
  args: string[],
): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npx', ['ng', 'generate', ...args]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}
