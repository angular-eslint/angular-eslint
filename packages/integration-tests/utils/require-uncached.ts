/**
 * It's important in our snapshot assertions which target modules on disk
 * that we are always grabbing a fresh copy of the file. Node's built-in
 * caching mechanism can get in the way of that so this utility helps
 * us work around the potential issue.
 *
 * (Having the return type as any makes it much easier to work with).
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function requireUncached(modulePath: string): any {
  delete require.cache[require.resolve(modulePath)];
  return require(modulePath);
}
