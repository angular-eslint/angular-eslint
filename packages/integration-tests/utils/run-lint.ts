/* eslint-disable @typescript-eslint/no-non-null-assertion */
import execa from 'execa';
import path from 'path';

import stripAnsi = require('strip-ansi');

const FIXTURES_DIR = path.join(__dirname, '../fixtures/');

function normalizeOutput(value: unknown): string {
  const ansiRemoved = stripAnsi(String(value));
  return (
    ansiRemoved
      .replace(
        new RegExp(`^${FIXTURES_DIR.replace(/\\/g, '\\\\')}(.*?)$`, 'gm'),
        (_, c1) => `__ROOT__/${c1.replace(/\\/g, '/')}`,
      )
      /**
       * This one is ridiculous...
       *
       * No matter what I try (including the use of `strip-ansi`), I do not seem to be
       * able to stop the inclusion of this extra `m` any other way (it is presumably
       * coming from an ANSI code becuase of its initial appearance in CI as [27m).
       * It happens locally and CI... but only exactly once.
       */
      .replace('mmax-len', 'max-len')
  );
}

export async function runLint(directory: string): Promise<string | undefined> {
  try {
    const subprocess = execa('npx', ['ng', 'lint'], {
      cwd: path.join(FIXTURES_DIR, directory),
    });

    /* eslint-disable @typescript-eslint/no-non-null-assertion */
    subprocess.stdout!.pipe(process.stdout);
    subprocess.stderr!.pipe(process.stderr);
    /* eslint-enable @typescript-eslint/no-non-null-assertion */

    const { stdout } = await subprocess;

    return normalizeOutput(stdout);
  } catch (error: any) {
    return normalizeOutput(error.stdout || error);
  }
}
