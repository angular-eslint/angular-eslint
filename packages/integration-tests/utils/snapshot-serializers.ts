function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function normalizeVersionOfPackage(str: string, pkg: string) {
  const regex = new RegExp(
    `("${escapeRegExp(pkg)}": "[~|^]\\d\\d?\\.)(\\d.\\d)"`,
  );
  return str.replace(regex, '$1X.X"');
}

const dependenciesToNormalize = [
  '@angular-devkit/build-angular',
  '@angular/cli',
  '@angular/compiler-cli',
  'ng-packagr',
  'typescript',
];

/**
 * Normalize dependencies controlled by the Angular CLI so that only the major version is explicitly
 * checked so that we can massively cut down on snapshot update noise across PRs.
 */
export const normalizeVersionsOfPackagesWeDoNotControl = {
  serialize(str: string) {
    for (const pkg of dependenciesToNormalize) {
      str = normalizeVersionOfPackage(str, pkg);
    }
    return str;
  },
  test(val: string) {
    // Only run if we think it's package.json contents
    return (
      val != null &&
      typeof val === 'string' &&
      dependenciesToNormalize.some((pkg) => val.includes(`"${pkg}"`))
    );
  },
};
