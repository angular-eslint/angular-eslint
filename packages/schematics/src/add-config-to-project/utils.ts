/**
 * Utils taken from various parts of Nx:
 * https://github.com/nrwl/nx
 *
 * Thanks, Nrwl folks!
 */
import { normalize } from '@angular-devkit/core';
import { Tree, SchematicContext, Rule } from '@angular-devkit/schematics';
import stripJsonComments from 'strip-json-comments';

/**
 * This method is specifically for reading JSON files in a Tree
 * @param host The host tree
 * @param path The path to the JSON file
 * @returns The JSON data in the file.
 */
function readJsonInTree<T = any>(host: Tree, path: string): T {
  if (!host.exists(path)) {
    throw new Error(`Cannot find ${path}`);
  }
  const contents = stripJsonComments(host.read(path)!.toString('utf-8'));
  try {
    return JSON.parse(contents);
  } catch (e) {
    throw new Error(`Cannot parse ${path}: ${e.message}`);
  }
}

function getWorkspacePath(host: Tree) {
  const possibleFiles = ['/workspace.json', '/angular.json', '/.angular.json'];
  return possibleFiles.filter((path) => host.exists(path))[0];
}

export function getProjectConfig(host: Tree, name: string): any {
  const workspaceJson = readJsonInTree(host, getWorkspacePath(host));
  const projectConfig = workspaceJson.projects[name];
  if (!projectConfig) {
    throw new Error(`Cannot find project '${name}'`);
  } else {
    return projectConfig;
  }
}

export function offsetFromRoot(fullPathToSourceDir: string): string {
  const parts = normalize(fullPathToSourceDir).split('/');
  let offset = '';
  for (let i = 0; i < parts.length; ++i) {
    offset += '../';
  }
  return offset;
}

function serializeJson(json: any): string {
  return `${JSON.stringify(json, null, 2)}\n`;
}

export function updateWorkspaceInTree<T = any, O = T>(
  callback: (json: T, context: SchematicContext, host: Tree) => O,
): Rule {
  return (host: Tree, context: SchematicContext): Tree => {
    const path = getWorkspacePath(host);
    host.overwrite(
      path,
      serializeJson(callback(readJsonInTree(host, path), context, host)),
    );
    return host;
  };
}
