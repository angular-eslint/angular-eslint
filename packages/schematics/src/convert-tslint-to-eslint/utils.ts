import type { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import * as assert from 'assert';
import { sortObjectByKeys } from '../utils';

export function updateArrPropAndRemoveDuplication(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json: Record<string, any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configBeingExtended: Record<string, any>,
  arrPropName: string,
  deleteIfUltimatelyEmpty: boolean,
): void {
  json[arrPropName] = json[arrPropName] || [];
  configBeingExtended[arrPropName] = configBeingExtended[arrPropName] || [];
  json[arrPropName] = json[arrPropName].filter(
    (extended: string) => !configBeingExtended[arrPropName].includes(extended),
  );
  json[arrPropName] = Array.from(new Set(json[arrPropName]));
  if (deleteIfUltimatelyEmpty && json[arrPropName].length === 0) {
    delete json[arrPropName];
  }
}

export function updateObjPropAndRemoveDuplication(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json: Record<string, any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configBeingExtended: Record<string, any>,
  objPropName: string,
  deleteIfUltimatelyEmpty: boolean,
): void {
  json[objPropName] = json[objPropName] || {};
  configBeingExtended[objPropName] = configBeingExtended[objPropName] || {};

  for (const [name, val] of Object.entries(json[objPropName])) {
    const valueOfSamePropInExtendedConfig =
      configBeingExtended[objPropName][name];

    try {
      assert.deepStrictEqual(val, valueOfSamePropInExtendedConfig);
      delete json[objPropName][name];
      // eslint-disable-next-line no-empty
    } catch {}
  }

  if (deleteIfUltimatelyEmpty && Object.keys(json[objPropName]).length === 0) {
    delete json[objPropName];
  }
}

export function ensureESLintPluginsAreInstalled(
  eslintPluginsToBeInstalled: string[],
): Rule {
  return (host: Tree, context: SchematicContext) => {
    if (!eslintPluginsToBeInstalled?.length) {
      return;
    }

    if (!host.exists('package.json')) {
      throw new Error(
        'Could not find a `package.json` file at the root of your workspace',
      );
    }

    const projectPackageJSON = (host.read('package.json') as Buffer).toString(
      'utf-8',
    );
    const json = JSON.parse(projectPackageJSON);
    json.devDependencies = json.devDependencies || {};

    const pluginsToInstall = [];

    for (const pluginName of eslintPluginsToBeInstalled) {
      if (
        !json.devDependencies[pluginName] &&
        !json.dependencies?.[pluginName]
      ) {
        json.devDependencies[pluginName] =
          // Hard code eslint-plugin-jsdoc version for now so that breaking rule change does not apply
          pluginName === 'eslint-plugin-jsdoc' ? '41.1.2' : 'latest';
        pluginsToInstall.push(pluginName);
      }
    }

    if (pluginsToInstall.length > 0) {
      context.logger.info(
        '\nINFO: To most closely match your tslint.json, the `latest` version of the following eslint plugin(s) have been installed:',
      );
      context.logger.info('\n  - ' + pluginsToInstall.join('\n  - '));
      context.logger.info(
        '\nPlease note, you may wish to pin these to a specific version number in your package.json, rather than leaving it open to `latest`.\n',
      );

      host.overwrite('package.json', JSON.stringify(json, null, 2));
      context.addTask(new NodePackageInstallTask());
    }

    return host;
  };
}

export function uninstallTSLintAndCodelyzer(): Rule {
  return (host: Tree, context: SchematicContext) => {
    if (!host.exists('package.json')) {
      throw new Error(
        'Could not find a `package.json` file at the root of your workspace',
      );
    }

    const projectPackageJSON = (host.read('package.json') as Buffer).toString(
      'utf-8',
    );
    const json = JSON.parse(projectPackageJSON);

    if (json.devDependencies) {
      delete json.devDependencies['tslint'];
      delete json.devDependencies['codelyzer'];
      json.devDependencies = sortObjectByKeys(json.devDependencies);
    }

    if (json.dependencies) {
      delete json.dependencies['tslint'];
      delete json.dependencies['codelyzer'];
      json.dependencies = sortObjectByKeys(json.dependencies);
    }

    host.overwrite('package.json', JSON.stringify(json, null, 2));
    context.addTask(new NodePackageInstallTask());

    return host;
  };
}
