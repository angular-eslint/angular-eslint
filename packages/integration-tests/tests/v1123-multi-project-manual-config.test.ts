import path from 'path';
import {
  FIXTURES_DIR,
  LONG_TIMEOUT_MS,
  runNgAdd,
  runNpmInstall,
} from '../utils/local-registry-process';
import { runLint } from '../utils/run-lint';

const fixtureDirectory = 'v1123-multi-project-manual-config';

describe(fixtureDirectory, () => {
  jest.setTimeout(LONG_TIMEOUT_MS);

  beforeEach(async () => {
    const cwd = process.cwd();
    process.chdir(path.join(FIXTURES_DIR, fixtureDirectory));
    await runNpmInstall();
    await runNgAdd();
    process.chdir(cwd);
  });

  it('it should produce the expected lint output', async () => {
    const lintOutput = await runLint(fixtureDirectory);
    expect(lintOutput).toMatchSnapshot();
  });
});
