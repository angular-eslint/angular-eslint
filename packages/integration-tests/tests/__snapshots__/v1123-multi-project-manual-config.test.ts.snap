// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v1123-multi-project-manual-config it should produce the expected lint output 1`] = `
"
Linting \\"v1123-multi-project-manual-config\\"...

__ROOT__/v1123-multi-project-manual-config/src/app/app.component.html
  1:31  error  Invalid binding syntax. Use [(expr)] instead  @angular-eslint/template/banana-in-box

__ROOT__/v1123-multi-project-manual-config/src/app/app.component.spec.ts
  19:6  error  Strings must use singlequote  @typescript-eslint/quotes

__ROOT__/v1123-multi-project-manual-config/src/app/app.component.ts
   2:1  error  'rxjs/Rx' import is restricted from being used. Please import directly from 'rxjs' instead                             no-restricted-imports
   5:1  error  Unexpected console statement                                                                                           no-console
  15:3  error  Lifecycle interface 'OnInit' should be implemented for method 'ngOnInit'. (https://angular.io/styleguide#style-09-01)  @angular-eslint/use-lifecycle-interface

__ROOT__/v1123-multi-project-manual-config/src/app/example.component.ts
   4:13  error  Strings must use singlequote                                                                                                          @typescript-eslint/quotes
  12:3   error  Use @Input rather than the \`inputs\` metadata property (https://angular.io/styleguide#style-05-12)                                     @angular-eslint/no-inputs-metadata-property
  13:3   error  Use @Output rather than the \`outputs\` metadata property (https://angular.io/styleguide#style-05-12)                                   @angular-eslint/no-outputs-metadata-property
  14:3   error  Use @HostBinding or @HostListener rather than the \`host\` metadata property (https://angular.io/styleguide#style-06-03)                @angular-eslint/no-host-metadata-property
  18:3   error  Output bindings, including aliases, should not be named \\"on\\", nor prefixed with it (https://angular.io/guide/styleguide#style-05-16)  @angular-eslint/no-output-on-prefix
   6:35  error  Invalid binding syntax. Use [(expr)] instead                                                                                          @angular-eslint/template/banana-in-box
   8:15  error  Invalid binding syntax. Use [(expr)] instead                                                                                          @angular-eslint/template/banana-in-box
   8:29  error  Invalid binding syntax. Use [(expr)] instead                                                                                          @angular-eslint/template/banana-in-box
   9:48  error  Invalid binding syntax. Use [(expr)] instead                                                                                          @angular-eslint/template/banana-in-box

__ROOT__/v1123-multi-project-manual-config/src/app/example.pipe.ts
  6:14  error  Pipes should implement \`PipeTransform\` interface  @angular-eslint/use-pipe-transform-interface

__ROOT__/v1123-multi-project-manual-config/src/app/multiple-inline-templates.page.ts
   4:13  error  Strings must use singlequote                  @typescript-eslint/quotes
  12:13  error  Strings must use singlequote                  @typescript-eslint/quotes
   6:35  error  Invalid binding syntax. Use [(expr)] instead  @angular-eslint/template/banana-in-box
  14:48  error  Invalid binding syntax. Use [(expr)] instead  @angular-eslint/template/banana-in-box

✖ 19 problems (19 errors, 0 warnings)
  12 errors and 0 warnings potentially fixable with the \`--fix\` option.


Linting \\"another-app\\"...

__ROOT__/v1123-multi-project-manual-config/projects/another-app/src/app/app.component.html
   12:1  error  This line has a length of 159. Maximum allowed is 140   max-len
  306:1  error  This line has a length of 532. Maximum allowed is 140   max-len
  313:1  error  This line has a length of 540. Maximum allowed is 140   max-len
  327:1  error  This line has a length of 219. Maximum allowed is 140   max-len
  328:1  error  This line has a length of 523. Maximum allowed is 140   max-len
  335:1  error  This line has a length of 151. Maximum allowed is 140   max-len
  336:1  error  This line has a length of 288. Maximum allowed is 140   max-len
  347:1  error  This line has a length of 205. Maximum allowed is 140   max-len
  351:1  error  This line has a length of 184. Maximum allowed is 140   max-len
  354:1  error  This line has a length of 220. Maximum allowed is 140   max-len
  358:1  error  This line has a length of 176. Maximum allowed is 140   max-len
  362:1  error  This line has a length of 435. Maximum allowed is 140   max-len
  366:1  error  This line has a length of 176. Maximum allowed is 140   max-len
  379:1  error  This line has a length of 167. Maximum allowed is 140   max-len
  385:1  error  This line has a length of 167. Maximum allowed is 140   max-len
  391:1  error  This line has a length of 167. Maximum allowed is 140   max-len
  397:1  error  This line has a length of 165. Maximum allowed is 140   max-len
  403:1  error  This line has a length of 165. Maximum allowed is 140   max-len
  409:1  error  This line has a length of 165. Maximum allowed is 140   max-len
  429:1  error  This line has a length of 188. Maximum allowed is 140   max-len
  430:1  error  This line has a length of 158. Maximum allowed is 140   max-len
  431:1  error  This line has a length of 165. Maximum allowed is 140   max-len
  432:1  error  This line has a length of 165. Maximum allowed is 140   max-len
  433:1  error  This line has a length of 162. Maximum allowed is 140   max-len
  440:1  error  This line has a length of 187. Maximum allowed is 140   max-len
  441:1  error  This line has a length of 159. Maximum allowed is 140   max-len
  442:1  error  This line has a length of 188. Maximum allowed is 140   max-len
  443:1  error  This line has a length of 174. Maximum allowed is 140   max-len
  450:1  error  This line has a length of 172. Maximum allowed is 140   max-len
  457:1  error  This line has a length of 190. Maximum allowed is 140   max-len
  458:1  error  This line has a length of 160. Maximum allowed is 140   max-len
  462:1  error  This line has a length of 217. Maximum allowed is 140   max-len
  466:1  error  This line has a length of 187. Maximum allowed is 140   max-len
  474:1  error  This line has a length of 190. Maximum allowed is 140   max-len
  475:1  error  This line has a length of 160. Maximum allowed is 140   max-len
  476:1  error  This line has a length of 410. Maximum allowed is 140   max-len
  483:1  error  This line has a length of 2884. Maximum allowed is 140  max-len
  509:1  error  This line has a length of 261. Maximum allowed is 140   max-len
  514:1  error  This line has a length of 230. Maximum allowed is 140   max-len
  518:1  error  This line has a length of 148. Maximum allowed is 140   max-len
  519:1  error  This line has a length of 1046. Maximum allowed is 140  max-len

__ROOT__/v1123-multi-project-manual-config/projects/another-app/src/app/app.component.spec.ts
  19:6  error  Strings must use singlequote  @typescript-eslint/quotes

✖ 42 problems (42 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.


Linting \\"horse-plants\\"...

All files pass linting.


Linting \\"allergies-fusion\\"...

All files pass linting.


Linting \\"breakfast-cat\\"...

All files pass linting.


Linting \\"prints-soda\\"...

All files pass linting.


Linting \\"ice-cream-monster\\"...

All files pass linting.


Linting \\"shoe-prints\\"...

All files pass linting.


Linting \\"websites-system\\"...

All files pass linting.


Linting \\"robot-websites\\"...

All files pass linting.
"
`;
